File: src/test/java/com/planit/stepdefs/CartSteps.java
package com.planit.stepdefs;

import com.planit.pages.CartPage;
import com.planit.pages.HomePage;
import com.planit.pages.ShopPage;
import com.planit.utils.DriverManager;
import com.planit.utils.PriceUtils;
import io.cucumber.java.en.And;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import org.openqa.selenium.WebDriver;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;

public class CartSteps {

    private WebDriver driver;
    private HomePage homePage;
    private ShopPage shopPage;
    private CartPage cartPage;

    private final Map<String, Double>  unitPrices    = new HashMap<>();
    private final Map<String, Integer> quantities    = new HashMap<>();
    /** Actual subtotals read from the cart page during individual subtotal assertions. */
    private final Map<String, Double>  actualSubtotals = new HashMap<>();

    private void init() {
        driver   = DriverManager.getDriver();
        homePage = new HomePage(driver);
        shopPage = new ShopPage(driver);
        cartPage = new CartPage(driver);
    }

    // NOTE: "I am on the home page" is already defined in ContactSteps.
    // Cucumber shares step definitions across classes, so no duplicate definition needed here.
    // However we need to re-init our page references after the Given step runs.
    // We handle that by calling init() lazily in each step that needs it.

    @When("I navigate to the shop page")
    public void iNavigateToTheShopPage() {
        driver   = DriverManager.getDriver();
        homePage = new HomePage(driver);
        shopPage = homePage.goToShop();
        cartPage = new CartPage(driver);
    }

    @And("I add {int} of {string} to the cart")
    public void iAddOfToTheCart(int quantity, String productName) {
        if (shopPage == null) {
            shopPage = new ShopPage(DriverManager.getDriver());
        }
        shopPage.addToCart(productName, quantity);
        quantities.put(productName, quantity);
    }

    @And("I navigate to the cart page")
    public void iNavigateToTheCartPage() {
        if (shopPage == null) {
            shopPage = new ShopPage(DriverManager.getDriver());
        }
        cartPage = shopPage.goToCart();
    }

    @Then("the subtotal for {string} should equal unit price multiplied by {int}")
    public void theSubtotalForShouldEqualUnitPriceMultipliedBy(String productName, int multiplier) {
        if (cartPage == null) {
            cartPage = new CartPage(DriverManager.getDriver());
        }
        double unitPrice        = cartPage.getUnitPrice(productName);
        unitPrices.put(productName, unitPrice);
        double expectedSubtotal = PriceUtils.round2dp(unitPrice * multiplier);
        double actualSubtotal   = PriceUtils.round2dp(cartPage.getSubtotal(productName));

        assertThat(actualSubtotal)
                .as("Subtotal for '%s' should be %.2f (unit price %.2f × %d)",
                        productName, expectedSubtotal, unitPrice, multiplier)
                .isEqualTo(expectedSubtotal);
    }

    @And("the grand total should equal the sum of all subtotals")
    public void theGrandTotalShouldEqualTheSumOfAllSubtotals() {
        if (cartPage == null) {
            cartPage = new CartPage(DriverManager.getDriver());
        }
        List<CartPage.CartRow> rows = cartPage.getAllRows();
        double computedTotal = rows.stream()
                .mapToDouble(row -> row.subtotal)
                .sum();

        double grandTotal = cartPage.getGrandTotal();

        assertThat(PriceUtils.round2dp(grandTotal))
                .as("Grand total %.2f should equal sum of subtotals %.2f", grandTotal, computedTotal)
                .isEqualTo(PriceUtils.round2dp(computedTotal));
    }
}

File: PlanitAutomation/StepDefinitions/CartSteps.cs
using FluentAssertions;
using PlanitAutomation.Pages;
using PlanitAutomation.Utils;
using Reqnroll;

namespace PlanitAutomation.StepDefinitions;

/// <summary>
/// Step definitions for Cart scenario (TC3).
/// </summary>
[Binding]
public sealed class CartSteps(ScenarioContext scenarioContext)
{
    // ── Page-object references ─────────────────────────────────────────────────

    private BrowserManager BrowserManager =>
        scenarioContext.Get<BrowserManager>();

    private HomePage?  _homePage;
    private ShopPage?  _shopPage;
    private CartPage?  _cartPage;

    // ── Navigation ─────────────────────────────────────────────────────────────

    [When("I navigate to the shop page")]
    public async Task INavigateToTheShopPage()
    {
        _homePage ??= new HomePage(BrowserManager.Page);
        _shopPage   = await _homePage.GoToShopAsync();
    }

    [When("I navigate to the cart page")]
    public async Task INavigateToTheCartPage()
    {
        _shopPage ??= new ShopPage(BrowserManager.Page);
        _cartPage   = await _shopPage.GoToCartAsync();
    }

    // ── Cart actions ───────────────────────────────────────────────────────────

    [When(@"I add (\d+) of ""(.*)"" to the cart")]
    public async Task IAddOfToTheCart(int quantity, string productName)
    {
        _shopPage ??= new ShopPage(BrowserManager.Page);
        await _shopPage.AddToCartAsync(productName, quantity);
    }

    // ── Cart assertions ────────────────────────────────────────────────────────

    [Then(@"the subtotal for ""(.*)"" should equal unit price multiplied by (\d+)")]
    public async Task TheSubtotalForShouldEqualUnitPriceMultipliedBy(string productName, int multiplier)
    {
        _cartPage ??= new CartPage(BrowserManager.Page);

        var unitPrice        = await _cartPage.GetUnitPriceAsync(productName);
        var expectedSubtotal = PriceUtils.Round2Dp(unitPrice * multiplier);
        var actualSubtotal   = PriceUtils.Round2Dp(await _cartPage.GetSubtotalAsync(productName));

        actualSubtotal.Should().Be(expectedSubtotal,
            $"subtotal for '{productName}' should be {expectedSubtotal:F2} " +
            $"(unit price {unitPrice:F2} × {multiplier})");
    }

    [Then("the grand total should equal the sum of all subtotals")]
    public async Task TheGrandTotalShouldEqualTheSumOfAllSubtotals()
    {
        _cartPage ??= new CartPage(BrowserManager.Page);

        var rows         = await _cartPage.GetAllRowsAsync();
        var computedSum  = rows.Sum(r => r.Subtotal);
        var grandTotal   = await _cartPage.GetGrandTotalAsync();

        PriceUtils.Round2Dp(grandTotal).Should().Be(
            PriceUtils.Round2Dp(computedSum),
            $"grand total {grandTotal:F2} should equal sum of subtotals {computedSum:F2}");
    }
}
